Lib/compression/zstd/_zstdfile.py

```diff
@@ -0,0 +1,378 @@
+        given as "rb", "wb", "xb" and "ab" respectively.
+
+        Parameters
+        level: The compression level to use, defaults to ZSTD_CLEVEL_DEFAULT. Note,
+            in read mode (decompression), compression level is not supported.
+        options: A dict object, containing advanced compression
+            parameters.
+        zstd_dict: A ZstdDict object, pre-trained dictionary for compression /
+            decompression.
+        """
+        self._fp = None
+        self._closefp = False
+        self._mode = _MODE_CLOSED
+
+        # Read or write mode
+        if mode in ("r", "rb"):
+            if not isinstance(options, (type(None), dict)):
+                raise TypeError(
+                    (
+                        "In read mode (decompression), options argument "
+                        "should be a dict object, that represents decompression "
+                        "options."
+                    )
+                )
+            if level is not None:
+                raise TypeError("level argument should only be passed when writing.")
+            mode_code = _MODE_READ
+        elif mode in ("w", "wb", "a", "ab", "x", "xb"):
+            if not isinstance(level, (type(None), int)):
+                raise TypeError("level argument should be an int object.")
+            if not isinstance(options, (type(None), dict)):
+                raise TypeError("options argument should be an dict object.")
+            mode_code = _MODE_WRITE
+            self._compressor = ZstdCompressor(
+                level=level, options=options, zstd_dict=zstd_dict
+            )
+            self._pos = 0
+        else:
+            raise ValueError(f"Invalid mode: {mode!r}")
+
+        # File object
+        if isinstance(filename, (str, bytes, PathLike)):
+            if "b" not in mode:
+                mode += "b"
+            self._fp = builtins.open(filename, mode)
+            self._closefp = True
+        elif hasattr(filename, "read") or hasattr(filename, "write"):
+            self._fp = filename
+        else:
+            raise TypeError("filename must be a str, bytes, file or PathLike object")
+        self._mode = mode_code
+
+        if self._mode == _MODE_READ:
+            raw = self._READER_CLASS(
+                self._fp,
+                ZstdDecompressor,
+                trailing_error=ZstdError,
+                zstd_dict=zstd_dict,
+                options=options,
+            )
+            self._buffer = io.BufferedReader(raw)
+
+    def close(self):
+        """Flush and close the file.
+
+        May be called more than once without error. Once the file is
+        closed, any other operation on it will raise a ValueError.
+        """
+        # Nop if already closed
+        if self._fp is None:
+            return
+        try:
+            if self._mode == _MODE_READ:
+                if hasattr(self, "_buffer") and self._buffer:
```

Comment:
we always have `self._buffer` if mode is MODE_READ, the guard isn't needed.
```suggestion
                if self._buffer:
```

If you want to keep the guard, simplify it to `getattr(self, '_buffer', None)`, but also ensure `self._compressor` below is guarded.

Reply:
When I remove the guard, `FileTestCase.test_init_bad_check` fails:

```
test_init_bad_check (test.test_zstd.test_core.FileTestCase.test_init_bad_check) ... Warning -- Unraisable exception
Exception ignored while finalizing file <compression.zstd._zstdfile.ZstdFile object at 0x7fcac7a7cc30>:
Traceback (most recent call last):
  File "/home/emma/cpython/Lib/compression/zstd/_zstdfile.py", line 117, in close
    self._buffer.close()
AttributeError: 'ZstdFile' object has no attribute '_buffer'
```

The getattr replacement is brilliant though, I will definitely use that.

Reply:
Right I just thought through when that will be the case: if `__init__` fails before `self._buffer` is initialized, then it will have this issue. I'll add a defensive `self._buffer = None` to the beginning to ensure that case won't happen (I think I am going to leave the check in `close()` to be defensive though).